src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java

```diff
@@ -193,52 +182,55 @@ private ConditionalDecoder(String conditionalsPropertyKey, Codec<List<ICondition
-        }
-
-        private <T> DataResult<Optional<A>> decodeInner(DynamicOps<T> ops, T input) {
             if (ops.compressMaps()) {
-                final DataResult<T> valueResult = ops.get(input, valuePropertyKey);
-                if (valueResult.error().map(DataResult.PartialResult::message).isPresent()) {
-                    return DataResult.error(() -> "Input does not have a \"value\" entry: " + input);
-                }
-                return valueResult.flatMap(value -> innerCodec.parse(ops, value).map(Optional::of));
-            }
-            if (innerCodec instanceof MapCodec.MapCodecCodec<A> mapCodec) {
-                return mapCodec.decode(ops, input).map(Pair::getFirst).map(Optional::of);
-            }
-
-            final DataResult<T> valueDataCarrierResult = ops.get(input, valuePropertyKey);
-
-            if (valueDataCarrierResult.error().isPresent()) {
-                return innerCodec.decode(ops, input).map(Pair::getFirst).map(Optional::of);
+                // Compressing ops are not supported at the moment because they require special handling.
+                return DataResult.error(() -> "Cannot use ConditionalCodec with compressing DynamicOps");
             }
 
-            return valueDataCarrierResult.flatMap(valueDataCarrier -> innerCodec.decode(ops, valueDataCarrier).map(Pair::getFirst).map(Optional::of));
-        }
+            return ops.getMap(input).map(inputMap -> {
+                final T conditionsDataCarrier = inputMap.get(conditionalsPropertyKey);
+                if (conditionsDataCarrier == null) {
+                    // No conditions, forward to inner codec
+                    return innerCodec.decode(ops, input).map(result -> result.mapFirst(carrier -> Optional.of(new WithConditions<>(carrier))));
+                }
 
-        @Override
-        public String toString() {
-            return "Conditional[inner=" + innerCodec + ", conditionalsKey=" + conditionalsPropertyKey + "]";
+                return conditionsCodec.decode(ops, conditionsDataCarrier).flatMap(conditionsCarrier -> {
+                    final List<ICondition> conditions = conditionsCarrier.getFirst();
+                    final DataResult<Pair<ICondition.IContext, T>> contextDataResult = contextCodec.decode(ops, ops.emptyMap());
+
+                    return contextDataResult.flatMap(contextCarrier -> {
+                        final ICondition.IContext context = contextCarrier.getFirst();
+
+                        final boolean conditionsMatch = conditions.stream().allMatch(c -> c.test(context));
+                        if (!conditionsMatch)
+                            return DataResult.error(() -> "Conditions did not match", Pair.of(Optional.empty(), input));
+
+                        DataResult<Pair<A, T>> innerDecodeResult;
+
+                        T valueDataCarrier = inputMap.get(valuePropertyKey);
+                        if (valueDataCarrier != null && inputMap.entries().count() == 2) {
```

Comment:
This isn't the proper logic - this would fail if, for instance, someone wrapped a `someCodec.fieldOf("value").codec()` in a conditional codec. If there's only a single value field, you have to try decoding with *both* the inner decoding directly of the value field and removing the conditional field and decoding that.

Reply:
Fixed - now `neoforge:value` is used as the value key.